src/qiskit_qasm3_import/state.py

```diff
@@ -0,0 +1,233 @@
+            raise_from_node(
+                node,
+                "Physical qubit referenced in virtual addressing mode. Mixing modes not currently supported.",
+            )
+        self._state = self.PHYSICAL
+
+    def set_virtual_mode(self, node):
+        """Set the addressing mode to virtual. On success return `True`, otherwise raise an exception."""
+        if self._state == self.PHYSICAL:
+            raise_from_node(
+                node,
+                "Virtual qubit declared in physical addressing mode. Mixing modes not currently supported.",
+            )
+        self._state = self.VIRTUAL
+
+
+class SymbolTable:
+    __slots__ = ("_local_table", "_global_table", "_builtin_table")
+
+    def __init__(self):
+        self._global_table = {}
+        self._builtin_table = _BUILTINS.copy()
+        self._local_table = {}  # For everything else
+
+    def gate_scope_copy(self):
+        """Return a copy of the symbol table for use in the lexical scope of a gate definition.
+
+        The target (returned) symbol table contains: all builtin symbols, no symbols from a surrounding local scope,
+        all symbols from the global scope referring to gates gates, all symbols from the global scope
+        that are marked constant and refer to numeric data.
+        """
+        # pylint: disable=protected-access
+        out = SymbolTable.__new__(SymbolTable)
+        out._local_table = {}
+        out._builtin_table = self._builtin_table
+        out._global_table = {}
+        for name, item in self._global_table.items():
+            if (
+                isinstance(item.type, types.Gate)
+                or isinstance(
+                    item.type, (types.Int, types.Uint, types.Float, types.Angle, types.Duration)
+                )
+                and item.type.const
+            ):
+                out._global_table[name] = item  # TODO: use insert
+        return out
+
+    def local_scope_copy(self):
+        """Return a copy of the symbol table for use with a new local scope.
+
+
+        Local variables created in the target (returned) symbol table will not
+        appear in the source symbol table. Thus, these local variables are
+        discarded and not visible upon returning to the surrounding scope. In
+        contrast, changes to data referred to by global symbols will be visible
+        in the surrounding scope.
+        """
+        # pylint: disable=protected-access
+        out = SymbolTable.__new__(SymbolTable)
+        out._local_table = self._local_table.copy()
+        out._global_table = self._global_table
+        out._builtin_table = self._builtin_table
+
+        return out
+
+    def insert(self, symbol: Union[List, Symbol]):
+        """Insert a `Symbol` (or each of a list thereof) into the symbol table."""
+        if isinstance(symbol, list):
+            for sym in symbol:
+                self._insert(sym)
+        else:
+            self._insert(symbol)
```

Comment:
I don't know the most satisfactory way to handle this in Python. I wanted to avoid the following, which I often see in Python and Qiskit  in particular
```python
def insert(self, symbol: Union[List, Symbol]):
    if not isinstance(symbol, list):
        symbol = [symbol]
    for sym in symbol:
        if symbol.scope == Scope.GLOBAL:
            self._global_table[symbol.name] = symbol
        else:
            self._local_table[symbol.name] = symbol
```
You can patch this up to detect tuples, etc.

But in fact I think I just have `insert` support single symbols. All calls that send more than one symbol come through a single function. I pretty sure I can put the loop there (`tests/test_values.py` , `_make_context`).

Reply:
Yeah, I meant I expect the caller to handle simple loops like this (e.g. `for symbols in symbols: table.insert(symbol)`).  There's a reason to do autovectorisation/broadcasting in something like Numpy, but not in raw Python imo - having the explicit loops are clearer at the call site, which is important for code clarity.

If you were going to do the autobroadcast, I'm not sure why you want to avoid the form you've given there - the code reuse is much simpler without needing an inner function (except in this case it should be `if isinstance(symbol, Symbol): symbol = [symbol]` so it worked with arbitrary iterables).